File: sdk/python/src/tinyhumans/api/invite.py
from __future__ import annotations

from typing import Any

from ..http import Json
from ._base import ApiNamespace

__all__ = ["InviteApi"]


class InviteApi(ApiNamespace):
    """Personal invite codes, redemption, and status checks."""

    def list_my_codes(self, **kwargs: Any) -> Json:
        """List the current user's invite codes with usage info."""
        return self._http.get("/invite/my-codes", **kwargs)

    def redeem_invite(self, body: dict[str, Any], **kwargs: Any) -> Json:
        """Redeem an invite code."""
        return self._http.post("/invite/redeem", body=body, **kwargs)

    def get_invite_status(
        self, query: dict[str, Any] | None = None, **kwargs: Any
    ) -> Json:
        """Check if an invite code is valid and available."""
        merged = self._merge_query(query, kwargs.pop("query", None))
        return self._http.get("/invite/status", query=merged, **kwargs)

File: src/api/invite.rs
//! Personal invite codes, redemption, and status checks.

use reqwest::Method;
use serde_json::Value;

use super::types::CodeRequest;
use crate::{Error, HttpClient, QueryParam};

/// Typed client for the `/invite/*` routes.
pub struct InviteApi<'a> {
    http: &'a HttpClient,
}

impl<'a> InviteApi<'a> {
    pub fn new(http: &'a HttpClient) -> Self {
        Self { http }
    }

    /// List the current user's invite codes with usage info.
    pub async fn list_my_codes(&self) -> Result<Value, Error> {
        self.http
            .send(Method::GET, "/invite/my-codes", &[], None, true)
            .await
    }

    /// Redeem an invite code.
    pub async fn redeem_invite(&self, request: &CodeRequest) -> Result<Value, Error> {
        let body = serde_json::to_value(request).expect("invite request is serializable");
        self.http
            .send(Method::POST, "/invite/redeem", &[], Some(&body), true)
            .await
    }

    /// Check if an invite code is valid and available.
    pub async fn get_invite_status(&self, query: &[QueryParam]) -> Result<Value, Error> {
        self.http
            .send(Method::GET, "/invite/status", query, None, true)
            .await
    }
}
